joomla/joomla-cms
# [4.0] Child templates

Pull Request is the foundation for #30149

### Summary of Changes
Introduces Inheritable templates solving numerous problem with the current state of templates. Read more on the linked issue above

This PR is complete in the sense that contains all the changes (even the upgrade path from J3) to introduce the new mode but without introducing the GUI and neither is moving any of the templates to the new mode. These can be done in following PR's


### Testing Instructions

Do not try to test this with patch tester, it won't work.
Install the PR's installable, you can d/l it here (at the bottom of this page)
If you want to use git command lines or the desktop app make sure that you'll do a fresh install or apply the updates (sql files)

Make sure you have the browser's console open so you can observe any missing assets.

If this went smooth, then it's a good sign as the installation is a seperate application with it's own template and although this PR is not touching the installation template everything works as before. B/C check 1

### Front End step 1

Go to the front end and vist the following urls:

- /index.php/blog
- /index.php/blog/3-welcome-to-your-blog
- /index.php/author-login (login here)
- /index.php?option=com_config&view=modules&id=109&Itemid=104 (or just try to edit the top left module)

Everything should be fine, no console logs or missinng icons, etc. B/C check 2

### Front End step 2

Download this [cassiopeia_overrides_v2.zip](https://github.com/joomla/joomla-cms/files/4977969/cassiopeia_overrides_v2.zip) and extract the contents and copy them in the folder `templates/cassiopeia/html`. Revisit the pages of the previous step. You should see in the pages some plain text (that will also break a bit the design) `Legacy Component view override`. You have to look for `Component`, `Module`, `Plugin` and `JLayout`. If so then the cascading of the templating is working correctly, in the previous step there was no overrides in this one the overrides were treated as top priority. Also cassiopeia has some overrides for the static assets so that was already tested (if you want you might open a normal 4.0 version in another winndow and check the number of assets per page between the 2). B/C check 3

### Front End step 3

Install the first new mode template (basically cassiopeia with another name: siteparent) from here [siteparent_v2.zip](https://github.com/joomla/joomla-cms/files/4977972/siteparent_v2.zip)


Go to the admin and make the siteparent template default and also assign all the menus to this template!!! Go to the front end and repeat the Front End step 1 and confirm that everything is as before. check 4

### Front End step 4

Go with your filemanager to `templates/siteparent` and copy the contents os the folder `xhtml` to the folder `html`. Redo the ritual of Front End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 5

### Front End step 5

Although the child templates should be created inside joomla's env for testing purposes we will do it using the installer, so install this: [sitechild_v2.zip](https://github.com/joomla/joomla-cms/files/4977974/sitechild_v2.zip)



Go to the admin and make the sitechild template default and also assign all the menus to this template!!! Go to the front end and repeat the Front End step 1 and confirm that everything is as before. check 6

### Front End step 6

Go with your filemanager to `templates/sitechild` and rename the folder `xhtml` to `html`. Redo the ritual of Front End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 7

### Back End step 1

Go to the addinistrator end and vist the following urls:

- /administrator/index.php
- /administrator/index.php?option=com_content&view=articles
- /administrator/index.php?option=com_installer&view=install

Everything should be fine, no console logs or missinng icons, etc. B/C check 8

### Back End step 2

Download this [atum_overrides_v2.zip](https://github.com/joomla/joomla-cms/files/4977975/atum_overrides_v2.zip) and extract the contents and copy them in the folder `administrator/templates/atum/html`. Revisit the pages of the previous step. You should see in the pages some plain text (that will also break a bit the design) `Legacy Component view override`. You have to look for `Component`, `Module`, `Plugin` and `JLayout`. If so then the cascading of the templating is working correctly, in the previous step there was no overrides in this one the overrides were treated as top priority. Also cassiopeia has some overrides for the static assets so that was already tested (if you want you might open a normal 4.0 version in another window and check the number of assets per page between the 2). B/C check 9

### Back End step 3

Install the new mode template (basically atum with another name: adminparent) from here 
[adminparent_v2.zip](https://github.com/joomla/joomla-cms/files/4977976/adminparent_v2.zip)


Go to the administrator template styles and make the adminparent template the default one. Repeat the Back End step 1 and confirm that everything is as before. check 10

### Back End step 4

Go with your filemanager to `administrator/templates/adminparent` and copy the contents os the folder `xhtml` to the folder `html`. Redo the dance of the Back End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 11

### Back End step 5

Althouth the child templates should be created inside joomla's env for testing purposes we will do it using the installer, so install this: 
[adminchild_v2.zip](https://github.com/joomla/joomla-cms/files/4977978/adminchild_v2.zip)



Go to the admin and make the adminchild template the default. Repeat the Back End step 1 and confirm that everything is as before. check 12

### Back End step 6

Go with your filemanager to `administrator/templates/adminchild` and rename the folder `xhtml` to `html`. Redo the ritual of the Back End step 1 and confirm that all the parts `Component`, `Module`, `Plugin` and `JLayout` are there. check 13

### Final checks

Check your db and the table template_styles, you should have

- autum and cassiopeia with parent `0` and inherits ``
- from the new ones the parent ones with parent `1` and inherits ``
- and the child ones ones with parent `0` and inherits `siteparent` and `siteadmin` respectfully

Finally someone needs to confirm that the update sql files are all that is needed here

### And the last one

You have probably saw some missing strings when you were trying to edit any of the template styles. This is expected as I forgot to include any languages in the parent template, but this is quite easy to check. You are already in the adminchild template so all you have to do is create a folder `language/en-GB` in the **adminparent** template and then copy the tmpl_atum\*.ini to this folder and rename them to tmpl_adminparent.ini and tmpl_adminparent.sys.ini. Check again the strings are back. Then make the adminparent the default template and check for any changes (readable text should remain readable). You can repeat this for the front end templates (although the function is common for both).


### Documentation Changes Required
For the devs the new parts are 2:
- manifest **requires** a tag `<inheritable>1</inheritable >` to enable a template in the new mode
- **all** static assets **and also any static asset override** should be done in the `media/( administrator || site )/templateName`. Folders like css/js/images inside the above path will act exactly the same as they did in the old `templates/name/( css || js || images )`
